### ISSUE TITLE
TagField Performance with large number of tags

### ISSUE THREAD

**Issue by caffeineinc** [silverstripe/silverstripe-tagfield#115]
`TagField` is used in SilverStripe Blog to reference tags, and autocomplete the list. However with a large number of total tags (~12000) Performance in the CMS is hindered due to a parent class hydrating the entire tag list for the form. 

`TagField` is a subclass of `DropdownField`, and `SelectField` and all descendants of FormFields must call `parent::__construct()`. Unfortunately this means that `SelectField::getListMap()` is called from setSource in the constructor of SelectField which transform the `HasManyRelation` (set as the source) and hydrates it into an array (Effectively returning all 12000 tags, and storing them locally in that object). 

For a SelectField that makes sense as you need the source to determine if a value should be allowed. `TagField` handles both states, both with and without source items.  If lazy load is used, the source shouldn't need the reference hydrated as it submits a request for this from the frontend (and a check on submission). 

The problem is this happens during construction, and lazy load could be applied at a later time, thus there's no easy way of assuming it other than using `sourceList` over the parent `source` - It makes sense that these objects should only be hydrated when actually used to avoid performance problems.

Are there any other suggestions towards resolving that problem? 

The fix would be changing the constructor to send an empty source array:  instead of the source
```
public function __construct($name, $title = '', $source = [], $value = null, $titleField = 'Title')
    {
        $this->setSourceList($source);
        $this->setTitleField($titleField);

//        parent::__construct($name, $title, $source, $value);

        parent::__construct($name, $title, [], $value);
    }
```

**Comment by caffeineinc** [comment]
`It works` 👍  haha 

Unfortunately i don't think a change would fit into semver, as the only way to change it would be omitting the source from the parent, and that would potentially break other behaviour that relies on the source. 

The component would have to know about lazy loading at instantiation, which means changing the interface or maybe splitting the class into a two components (lazyload tagfield and non-lazy tagfield). 

Happy to provide a PR but would like to establish direction before doing so.

**Comment by ScopeyNZ** [comment]
I suppose you could limit the breakage by extending `getSource`:

```
public function getSource()
{
    if (!$this->source) {
        $this->setSource($this->getSourceList());
    }
    return $this->source;
}
```

But you're right - anyone relying on `->source` would be affected.

**Comment by caffeineinc** [comment]
Yeah, further analysis also shows `TagField` mapping the entire `sourceList` for flagging and displaying the `checked` items  (it renders the entire tree :/) 

Since that too has some performance issues on render, it might be worth considering a seperate class for lazyload. (It should only know about the reference it requires for render, everything else should be lazyloaded).

**Comment by robbieaverill** [comment]
To be fair, `$this->source` is not part of our public API. Consumers should be using `$this->getSource()` instead. If we can ensure that still returns a DataList then it may not break BC

**Comment by caffeineinc** [comment]
OK, here's the attempt then: Significant improvements for all TagFields using this.

**Comment by tractorcow** [comment]
Concur with @ScopeyNZ looks like the best solution.

**Comment by robbieaverill** [comment]
Fixed with #117
